src/settings.txt
import {moduleName, MeistertoolsUtil} from "../meistertools.js";
import zutatenNsc from "../config/nsc-factory.config.zutaten.js";
import {NscFactory} from "./nsc-factory.js";
import {SceneParser} from "./scene-parser.js";
import {MeistertoolsRarity} from "./rarity.js";
import {Scenes} from "./scenes.js";

export const settingsCategories = [
    {key: "nsc-factory", default: NscFactory.defaultSettings},
    {key: "scenes", default: Scenes.defaultSettings},
    {key: "locations", default: SceneParser.defaultSettings},
    {key: "general", default: {showSettings: true}},
]


class MeisterApplication extends FormApplication {
    isOpen = false

    toggle() {
        if (this.isOpen)
            this.close()
        else
            this.render(true)
    }

    close() {
        this.isOpen = false
        super.close()
    }

    render(force) {
        this.isOpen = true
        super.render(force)
    }


    constructor() {
        super();
    }

    activateListeners(html) {
        super.activateListeners(html);
        MeistertoolsUtil.addDefaultListeners(html);
    }

}


export class MeistertoolsSettings extends MeisterApplication {
    constructor(initialTab = 'nsc') {
        super();
        this.initialTab = initialTab
        this.settings = {}
        for (let {key} of settingsCategories) {
            this.settings[key] = game.settings.get(moduleName, key)
        }

        this.selectOptions = {
            actorPacks: game.packs.filter(p => p.metadata.entity === 'Actor'),
            scenePacks: game.packs.filter(p => p.metadata.entity === 'Scene'),
            sceneFolders: game.folders.entities.filter(f => f.type === "Scene"),
            baseActors: [],
            playlists: game.playlists.entities.map(p => p.name),
        }

        this.zutatenInstalled = (game.modules.get("dsa5-meistertools-zutaten")?.active)
    }

    async _setSelectOptions() {
        const pack = game.packs.find(p => p.collection === this.settings["nsc-factory"]?.baseActorCollection)
        if (!pack?.index?.length)
            await pack?.getIndex()
        this.selectOptions.baseActors = pack?.index || []
    }


    static get defaultOptions() {
        return mergeObject(super.defaultOptions, {
            classes: ['meistertools'],
            popOut: true,
            width: 1020,
            height: 800,
            resizable: true,
            closeOnSubmit: false,
            submitOnChange: false,
            submitOnClose: false,
            template: `modules/${moduleName}/templates/settings.hbs`,
            id: 'meistertools.settings',
            title: 'MeisterTools Settings',
            tabs: [{navSelector: ".tabs", contentSelector: ".content", initial: "general"}]
        });
    }


    async getData() {
        await this._setSelectOptions()
        this.settings.locations.regions = this.settings.locations.regions.sort((a, b) => MeistertoolsUtil.strcmp(a.name, b.name))
        return {
            selectOptions: this.selectOptions,
            ...this.settings,
            zutatenInstalled: this.zutatenInstalled
        }
    }


    activateListeners(html) {
        super.activateListeners(html);
        MeistertoolsUtil.addDefaultListeners(html);
        html.find("button.open-app").click((event) => {
            const appName = $(event.currentTarget).attr("name")
            if (appName === 'scene-parser') {
                new SceneParser().render(true)
                this.close()
            }
            if (appName === 'entity-tagger') {
                new MeistertoolsRarity().render(true)
                this.close()
            }
        });
        html.find("button[type=reset]").click((event) => {
            const category = $(event.currentTarget).attr("data-category")
            this.settings[category] = game.settings.get(moduleName, category)
            this.render()
        });
        html.find("button[name=restore-settings]").click(async (event) => {
            const category = $(event.currentTarget).attr("data-category")
            await game.settings.set(moduleName, category, game.settings.settings.get(moduleName + '.' + category).default)
            this.settings[category] = await game.settings.get(moduleName, category)
            this.render()
        });
        html.find("button[name=install-zutaten]").click(async (event) => {
            const category = $(event.currentTarget).attr("data-category")
            if (category === "nsc-factory") {
                await game.settings.set(moduleName, category, zutatenNsc)
                this.settings[category] = await game.settings.get(moduleName, category)
                this.render()
            }
        });
        html.find("button[name=report-bug]").click(async () => {
            window.open("https://github.com/tyr0mancer/dsa5-meistertools/issues", "_blank")
        });


        html.find(".insert-region").click(event => {
            const name = html.find("input[name='tmp.region-name']").val()
            const key = html.find("input[name='tmp.region-key']").val()
            const category = html.find("select[name='tmp.region-category']").val()
            const biome = html.find("select[name='tmp.region-biome']").val()
            this.settings.locations.regions.push({name, key, category, biome})
            this.render()
        })

        html.find(".insert-entry").click(event => {
            const path = $(event.currentTarget).attr("data-path")?.split('.')
            if (!path) return
            if (!this.settings[path[0]][path[1]])
                this.settings[path[0]][path[1]] = []
            this.settings[path[0]][path[1]].push({})
            this.render()
        })

        html.find(".remove-entry").click(event => {
            const index = $(event.currentTarget).attr("data-index")
            const path = $(event.currentTarget).attr("data-path")?.split('.')
            if (!path || !this.settings[path[0]][path[1]]) return
            this.settings[path[0]][path[1]].splice(index, 1)
            this.render()
        })
    }


    _updateObject(event, formData) {
        formData = MeistertoolsUtil.expandObjectAndArray(formData)
        for (let {key: category} of settingsCategories) {
            if (!formData[category]) continue
            mergeObject(this.settings[category], formData[category])
            game.settings.set(moduleName, category, this.settings[category]);
        }
        Hooks.call(moduleName + ".update-settings", this.settings)
        this.close()
    }

}


modules/register-settings_.txt
import {moduleName} from "../meistertools.js";
import {ManageScenes} from "./manage-scenes.js";
import {CreateNSC, getRuleset} from "./create-nsc.js";

export async function registerSettings() {
    await game.settings.registerMenu(moduleName, "config-ui", {
        name: "DSA5 Meistertools",
        label: "Einstellungen",
        hint: "Alle Einstellungen der DSA5 Meistertools",
        icon: "fas fa-eye",
        type: MeistertoolsConfig,
        restricted: true
    });
    await game.settings.registerMenu(moduleName, "reset-settings", {
        name: "Initialize",
        label: "Initialisieren",
        type: InitializerForm,
        restricted: true
    });

    await game.settings.register(moduleName, "settings", {
        scope: "world",
        config: false,
        type: Object,
        default: MeistertoolsConfig.defaultSettings
    });
}


class InitializerForm extends FormApplication {

    render() {
        let content = `<h1>DSA5 Meistertools initialisieren</h1>`
        let buttons = {
            reset: {
                icon: '<i class="fa fa-check"></i>',
                label: 'zurück setzen',
                callback: html => {
                    game.settings.set(moduleName, 'settings', undefined)
                }
            },
            cancel: {
                icon: '<i class="fas fa-cancel"></i>',
                label: 'Abbruch',
                callback: html => {
                }
            }
        }

        if (game.modules.get('dsa5-homebrew')) {
            buttons['homebrew'] = {
                icon: '<i class="fa fa-check"></i>',
                label: 'Für dsa5-homebrew konfigurieren',
                callback: html => {
                    game.settings.set(moduleName, 'settings', {
                        "scenes": {
                            "activateDefault": true,
                            "updatePlaylist": true,
                            "defaultPlaylist": "",
                            "filterExisting": true,
                            "categories": [{
                                "name": "Theatre Of The Mind",
                                "folder": "TOTM",
                                "keywords": "Einkauf,Taverne,Tempel",
                                "packname": "dsa5-homebrew.scene-totm",
                                "addplayers": true,
                                "position": "untenlinks"
                            }, {
                                "name": "Battlemaps Wald",
                                "folder": "BM Wald",
                                "keywords": "mainroad,sideroad,well",
                                "packname": "dsa5-homebrew.scene-bm-forest",
                                "addplayers": false,
                                "position": "center",
                                "thumbFolder": "modules/dsa5-homebrew/images/battlemaps/forest-bm/thumbs",
                                "templateId": "gk4KBgTv0KjqkCKf"
                            }]
                        },
                        "nsc": {
                            "closeAfterGeneration": true,
                            "tokenImageFolder": "modules/dsa5-homebrew/images/actors/random-npc",
                            "defaultOrigin": "mittellande",
                            "defaultCulture": "garethisch",
                            "defaultProfession": "AENj6jLndvawKiHG",
                            "defaultProfessionName": "Bürger",
                            "storedPatterns": [],
                            "genderOptions": [{"key": "random", "icon": "fas fa-dice", "name": "zufall"}, {
                                "key": "w",
                                "icon": "fas fa-venus",
                                "name": "weiblich"
                            }, {"key": "m", "icon": "fas fa-mars", "name": "männlich"}],
                            "professionPack": "dsa5-homebrew.actor-archetypen",
                            "folder": "Zufalls NPC"
                        }
                    })
                }
            }
            content += `<p>DSA5 Homebrew wurde gefunden und kann für Meistertools konfiguriert werden!</p>`
        }
        new Dialog({title: 'Meistertools initialisieren', content, buttons, default: 'cancel'}).render(true)
    }
}


class MeistertoolsConfig extends FormApplication {
    constructor() {
        super();
    }

    static get defaultOptions() {
        return mergeObject(super.defaultOptions, {
            classes: ['form'],
            popOut: true,
            template: `modules/${moduleName}/templates/settings.html`,
            id: 'meistertools.settings',
            title: 'Meistertools Settings',
            tabs: [{navSelector: ".tabs", contentSelector: ".content", initial: "scenes"}]
        });
    }


    getData() {
        return game.settings.get(moduleName, 'settings');
    }

    activateListeners(html) {
        super.activateListeners(html);
    }

    _updateObject(event, formData) {
        const expandedData = expandObject(formData)
        mergeObject(game.settings.get(moduleName, 'settings'),expandedData)
        console.clear()
        console.log(expandedData)
        //game.settings.set(moduleName, 'settings', data);
    }

    static get defaultSettings() {
        return {
            scenes: ManageScenes.getDefaultSettings(),
            nsc: CreateNSC.getDefaultSettings(),
        }
    }
}


class MeistertoolsConfigOld extends FormApplication {

    static DEFAULT_SETTINGS() {
        return {
            scenes: ManageScenes.getDefaultSettings(),
            nsc: CreateNSC.getDefaultSettings(),
        }
    }

    constructor(object, options) {
        super(object, options);
        game.settings.sheet.close();

        const scenePacks = game.packs
            .filter(p => p.metadata.entity === 'Scene')
            .map(p => p.metadata.package + '.' + p.metadata.name)

        const actorPacks = game.packs
            .filter(p => p.metadata.entity === 'Actor')
            .map(p => p.metadata.package + '.' + p.metadata.name)

        let origins = []
        let cultures = []
        for (let archetype of getRuleset().archetypes)
            for (let origin of archetype.origins) {
                origins.push({key: origin.key, name: origin.name})
                if (origin.cultures !== undefined)
                    cultures = cultures.concat(origin.cultures)
            }

        this.dataObject = mergeObject({
            scenePacks,
            actorPacks,
            origins,
            cultures,
            playlists: game.playlists
        }, game.settings.get(moduleName, 'settings') || MeistertoolsConfig.DEFAULT_SETTINGS());
    }

    static get defaultOptions() {
        const options = super.defaultOptions;
        options.id = moduleName + '-config';
        options.title = 'DSA5 Meistertools Einstellungen';
        options.template = `modules/${moduleName}/templates/meistertools-config.html`;
        options.tabs = [{navSelector: ".tabs", contentSelector: ".content", initial: "scenes"}]
        options.closeOnSubmit = true;
        options.width = 400
        options.resizable = true
        return options;
    }

    getData() {
        return this.dataObject
    }

    render(force, context = {}) {
        return super.render(force, context);
    }


    activateListeners(html) {
        super.activateListeners(html);
        html.find('.insert-element').click(event => this._insertArrayElement(event));
        html.find('.delete-element').click(event => this._deleteArrayElement(event));

        html.find('.add-scene-pack').click(ev => {
            ev.preventDefault();
            if (this.dataObject.scenes.categories === undefined)
                this.dataObject.scenes.categories = []
            this.dataObject.scenes.categories.push({})
            this.render()
        });

        html.find("button[name='change-token-image-folder']").click(event => this._changeTokenImageFolder());


        html.find('.delete-scene-pack').click(ev => {
            ev.preventDefault();
            this.dataObject.scenes.categories.splice(ev.target.value, 1);
            this.render();
        });
    }

    _deleteArrayElement(event) {
        event.preventDefault();
        console.log(event.target.name)
        console.log(event.target.value)
        const path = event.target.name.split('.')
        this.dataObject[path[0]][path[1]].splice(event.target.value, 1);
        this.render()
    }

    _insertArrayElement(event) {
        event.preventDefault();
        //console.log(event.target.name)
        const path = event.target.name.split('.')
        if (this.dataObject[path[0]][path[1]] === undefined)
            this.dataObject[path[0]][path[1]] = []
        this.dataObject[path[0]][path[1]].push({})
        this.render()
    }


    // todo there GOTTA be a better way to do this dynamically with handlebars data-binding. maybe someday Ill learn how
    _updateObject(event, formData) {
        let updatedSettings = game.settings.get(moduleName, 'settings')
        const settingMainKeys = Object.keys(updatedSettings)
        const formDataKeys = Object.keys(formData)
        for (let mainKey of settingMainKeys) {
            let arrayData = {}
            for (let formDataKey of formDataKeys) {
                if (formDataKey.startsWith(mainKey)) {
                    let dataKey = formDataKey.substr(mainKey.length + 1)
                    let dataValue = formData[formDataKey]
                    let arrayIndex = dataKey.indexOf('_')
                    if (arrayIndex !== -1) {
                        let arrayName = dataKey.substr(0, arrayIndex);
                        if (arrayData[arrayName] === undefined)
                            arrayData[arrayName] = []
                        arrayData[arrayName].push(dataKey.substr(arrayName.length + 1))
                        continue
                    }
                    updatedSettings[mainKey][dataKey] = dataValue
                }
            }
            const arrayKeys = Object.keys(arrayData)
            for (let arrayKey of arrayKeys) {
                let arr = []
                for (let arrayElem of arrayData[arrayKey]) {
                    const elem = formData[mainKey + '.' + arrayKey + "_" + arrayElem]
                    if (!Array.isArray(elem)) {
                        if (arr.length === 0) arr = [{}]
                        arr[0][arrayElem] = elem
                    } else {
                        if (arr.length === 0)
                            for (let index = 0; index < elem.length; index++)
                                arr.push({})
                        for (let index = 0; index < elem.length; index++)
                            arr[index][arrayElem] = elem[index]
                    }
                }
                updatedSettings[mainKey][arrayKey] = arr
            }
        }
        game.settings.set(moduleName, 'settings', updatedSettings);
    }

    _setTokenImageFolder(path) {
        this.dataObject.nsc.tokenImageFolder = path
        this.render()
    }

    async _changeTokenImageFolder() {
        await new FilePicker({
            type: "image",
            current: this.dataObject.nsc.tokenImageFolder,
            callback: (imagePath) => this._setTokenImageFolder(imagePath),
        }).browse(this.dataObject.nsc.tokenImageFolder, {wildcard: true});
    }
}


src/settings.js
import {moduleName, MeistertoolsUtil} from "../meistertools.js";
import zutatenNsc from "../config/nsc-factory.config.zutaten.js";
import {NscFactory} from "./nsc-factory.js";
import {SceneParser} from "./scene-parser.js";
import {MeistertoolsRarity} from "./rarity.js";
import {Scenes} from "./scenes.js";

export const settingsCategories = [
    {key: "nsc-factory", default: NscFactory.defaultSettings},
    {key: "scenes", default: Scenes.defaultSettings},
    {key: "locations", default: SceneParser.defaultSettings},
    {key: "general", default: {showSettings: true}},
]


class MeisterApplication extends FormApplication {
    constructor() {
        super();
        this.isOpen = false
    }

    toggle() {
        if (this.isOpen)
            this.close()
        else
            this.render(true)
    }

    close() {
        this.isOpen = false
        super.close()
    }

    render(force) {
        this.isOpen = true
        super.render(force)
    }


    activateListeners(html) {
        super.activateListeners(html);
        MeistertoolsUtil.addDefaultListeners(html);
    }

}


export class MeistertoolsSettings extends MeisterApplication {
    constructor() {
        super();
        this.settings = {}
        for (let {key} of settingsCategories) {
            this.settings[key] = game.settings.get(moduleName, key)
        }

        this.selectOptions = {
            actorPacks: game.packs.filter(p => p.metadata.entity === 'Actor'),
            scenePacks: game.packs.filter(p => p.metadata.entity === 'Scene'),
            sceneFolders: game.folders.entities.filter(f => f.type === "Scene"),
            baseActors: [],
            playlists: game.playlists.entities.map(p => p.name),
        }

        this.zutatenInstalled = (game.modules.get("dsa5-meistertools-zutaten")?.active)
    }

    async _setSelectOptions() {
        const pack = game.packs.find(p => p.collection === this.settings["nsc-factory"]?.baseActorCollection)
        if (!pack?.index?.length)
            await pack?.getIndex()
        this.selectOptions.baseActors = pack?.index || []
    }


    static get defaultOptions() {
        return mergeObject(super.defaultOptions, {
            classes: ['meistertools'],
            popOut: true,
            width: 1020,
            height: 800,
            resizable: true,
            closeOnSubmit: false,
            submitOnChange: false,
            submitOnClose: false,
            template: `modules/${moduleName}/templates/settings.hbs`,
            id: 'meistertools.settings',
            title: 'MeisterTools Settings',
            tabs: [{navSelector: ".tabs", contentSelector: ".content", initial: "general"}]
        });
    }


    async getData() {
        await this._setSelectOptions()
        this.settings.locations.regions = this.settings.locations.regions.sort((a, b) => MeistertoolsUtil.strcmp(a.name, b.name))
        return {
            selectOptions: this.selectOptions,
            ...this.settings,
            zutatenInstalled: this.zutatenInstalled
        }
    }


    activateListeners(html) {
        super.activateListeners(html);
        MeistertoolsUtil.addDefaultListeners(html);
        html.find("button.open-app").click((event) => {
            const appName = $(event.currentTarget).attr("name")
            if (appName === 'scene-parser') {
                new SceneParser().render(true)
                this.close()
            }
            if (appName === 'entity-tagger') {
                new MeistertoolsRarity().render(true)
                this.close()
            }
        });
        html.find("button[type=reset]").click((event) => {
            const category = $(event.currentTarget).attr("data-category")
            this.settings[category] = game.settings.get(moduleName, category)
            this.render()
        });
        html.find("button[name=restore-settings]").click(async (event) => {
            const category = $(event.currentTarget).attr("data-category")
            await game.settings.set(moduleName, category, game.settings.settings.get(moduleName + '.' + category).default)
            this.settings[category] = await game.settings.get(moduleName, category)
            this.render()
        });
        html.find("button[name=install-zutaten]").click(async (event) => {
            const category = $(event.currentTarget).attr("data-category")
            if (category === "nsc-factory") {
                await game.settings.set(moduleName, category, zutatenNsc)
                this.settings[category] = await game.settings.get(moduleName, category)
                this.render()
            }
        });
        html.find("button[name=report-bug]").click(async () => {
            window.open("https://github.com/tyr0mancer/dsa5-meistertools/issues", "_blank")
        });


        html.find(".insert-region").click(event => {
            const name = html.find("input[name='tmp.region-name']").val()
            const key = html.find("input[name='tmp.region-key']").val()
            const category = html.find("select[name='tmp.region-category']").val()
            const biome = html.find("select[name='tmp.region-biome']").val()
            this.settings.locations.regions.push({name, key, category, biome})
            this.render()
        })

        html.find(".insert-entry").click(event => {
            const path = $(event.currentTarget).attr("data-path")?.split('.')
            if (!path) return
            if (!this.settings[path[0]][path[1]])
                this.settings[path[0]][path[1]] = []
            this.settings[path[0]][path[1]].push({})
            this.render()
        })

        html.find(".remove-entry").click(event => {
            const index = $(event.currentTarget).attr("data-index")
            const path = $(event.currentTarget).attr("data-path")?.split('.')
            if (!path || !this.settings[path[0]][path[1]]) return
            this.settings[path[0]][path[1]].splice(index, 1)
            this.render()
        })
    }


    _updateObject(event, formData) {
        formData = MeistertoolsUtil.expandObjectAndArray(formData)
        for (let {key: category} of settingsCategories) {
            if (!formData[category]) continue
            mergeObject(this.settings[category], formData[category])
            game.settings.set(moduleName, category, this.settings[category]);
        }
        Hooks.call(moduleName + ".update-settings", this.settings)
        this.close()
    }

}
